src/metabase/models/field_values.clj

```diff
@@ -106,36 +106,50 @@
-  [distinct-values]
-  ;; only consume enough values to determine whether the total length is > `total-max-length` -- if it is, we can stop
-  (let [total-length (reduce
-                      (fn [total-length v]
-                        (let [new-total (+ total-length (count (str v)))]
-                          (if (>= new-total total-max-length)
-                            (reduced new-total)
-                            new-total)))
-                      0
-                      distinct-values)]
-    (u/prog1 (<= total-length total-max-length)
-      (log/debug (trs "Field values total length is > {0}." total-max-length)
-                 (if <>
-                   (trs "FieldValues are allowed for this Field.")
-                   (trs "FieldValues are NOT allowed for this Field."))))))
+
+(defn take-by-length
+  "Returns a transducer that takes a collection and returns a collection where the total length of all items in list is less than `max-length`."
+  [max-length]
+  (fn [rf]
+    (let [current-length (volatile! 0)]
+      (fn
+        ([] (rf))
+        ([result]
+         (rf result))
+        ([result input]
+         (vswap! current-length + (count (str input)))
+         (if (< @current-length max-length)
+           (rf result input)
+           (reduced result)))))))
 
 (defn distinct-values
   "Fetch a sequence of distinct values for `field` that are below the [[total-max-length]] threshold. If the values are
-  past the threshold, this returns `nil`. (This function provides the values that normally get saved as a Field's
+  past the threshold, this returns a subset of possible values values where the total length of all items is less than [[total-max-length]].
+  It also returns a `has_more_values` flag, `has_more_values` = `true` when the returned values list is a subset of all possible values.
+
+  ;; (distinct-values (Field 1))
+  ;; ->  {:values          [1, 2, 3]
+          :has_more_values false}
+
+  (This function provides the values that normally get saved as a Field's
   FieldValues. You most likely should not be using this directly in code outside of this namespace, unless it's for a
   very specific reason, such as certain cases where we fetch ad-hoc FieldValues for GTAP-filtered Fields.)"
   [field]
   (classloader/require 'metabase.db.metadata-queries)
   (try
-    (let [values ((resolve 'metabase.db.metadata-queries/field-distinct-values) field)]
-      (when (values-less-than-total-max-length? values)
-        values))
+    (let [distinct-values ((resolve 'metabase.db.metadata-queries/field-distinct-values) field)
+          values          (into [] (take-by-length total-max-length) distinct-values)]
+      {:values          values
+       ;; has_more_values=true means the list of values we return is a subset of all possible values.
+       ;; has_more_values=true when:
+       ;; 1. the number of distinct values of this field exceeded [[metabase.db.metadata-queries/absolute-max-distinct-values-limit]]
```

Comment:
This line confused me at first when I read it, and tried to reconcile it with the `=` being used on L151.

My understanding is now: `metabase.db.metadata-queries/field-distinct-values` will return a maximum of `absolute-max-distinct-values-limit` fields. If the result is exactly equal to the limit, we err on the side of assuming there were more values we didn't read. Is that right? If so, could you clarify this somehow - maybe just add a comment explaining why `=` is being used in L151?

Just brainstorming now...we could change the query to have a limit of `absolute-max-distinct-values-limit + 1` but keep the original limit for the number of fields we actually store. Then we can use a `>` comparison and be very sure whether there are extra values in the underlying field. Maybe not worth the effort though.

Reply:
yea it's a bit tricky, I updated the docs to make it clearer. Could you have a second look?